File: Problem-1_Search_2D_Matrix.py
class Solution:
    def searchMatrix(self, matrix, target):
        
        if not matrix:
            # Reject invalid input, empty 1D list
            return False
        
        # Get height and width from input matrix
        m, n = len(matrix), len(matrix[0])
        
        if m * n == 0:
            # Reject invalid input, empty 2D list
            return False
        
        # Binary seach on head elements of each row
        lower, upper = 0, m-1
        while lower <= upper:
            
            mid = lower + (upper-lower)//2
            
            mid_row_head_value = matrix[mid][0]
            
            if mid_row_head_value > target:
                upper = mid-1
            elif mid_row_head_value < target:
                lower = mid+1
            else:
                return True

        # Binary seach on the possible row
        row_idx = upper
        lower, upper = 0, n-1
        
        while lower <= upper:
            
            mid = lower + (upper-lower)//2
            
            mid_value = matrix[row_idx][mid]
            
            if mid_value > target:
                upper = mid-1
            elif mid_value < target:
                lower = mid+1
            else:
                return True
            
        return False

#Implementing Search for 2D matrix problem using Binary search with time complexity O(logm + logn)

File: Problem1_Search2DMatrix.cs
// Approach 1: each row is sorted - if we compare target with last element in each row and figure out the desired row, then binary search that row; time complexity is O(m + log n)
// Current Approach: Visualize the 2D matrix as 1D array and perform binary search
// Time Complexity: O(log mn) where m is the number of rows, n is the number of cols
// Space Complexity: O(1)

public class Solution {
    public bool SearchMatrix(int[][] matrix, int target) {
        int rows = matrix.Length, cols = matrix[0].Length;
        
        int start = 0, end = rows*cols-1;
        
        while(start<=end){
            int mid = start + (end-start)/2;
            int r = mid/cols;
            int c = mid%cols;
            
            if(matrix[r][c] == target)
                return true;
            else if(target < matrix[r][c])
                end = mid-1;
            else
                start = mid+1;
        }
        
        return false;
    }
}